sponsorship_tracking/model/contracts.py

```diff
@@ -0,0 +1,332 @@
+        display_states = list()
+        for sds_state in sds_states:
+            sponsorship_count = self.search_count([
+                ('sds_state', '=', sds_state[0])])
+            if sponsorship_count:
+                display_states.append(sds_state)
+        return display_states, fold
+
+    @api.model
+    def _read_group_fill_results(self, domain, groupby,
+                                 remaining_groupbys, aggregated_fields,
+                                 read_group_result,
+                                 read_group_order=None):
+        """
+        The method seems to support grouping using m2o fields only,
+        while we want to group by a simple status field.
+        Hence the code below - it replaces simple status values
+        with (value, name) tuples.
+        """
+        if groupby == 'sds_state':
+            state_dict = dict(self._get_sds_states())
+            for result in read_group_result:
+                state = result[groupby]
+                result[groupby] = (state, state_dict.get(state))
+
+        return super(recurring_contract, self)._read_group_fill_results(
+            domain, groupby, remaining_groupbys, aggregated_fields,
+            read_group_result, read_group_order
+        )
+
+    _group_by_full = {
+        'sds_state': sds_kanban_groups,
+    }
+
+    ##########################################################################
+    #                            WORKFLOW METHODS                            #
+    ##########################################################################
+    @api.multi
+    def contract_validation(self):
+        wf_service = netsvc.LocalService('workflow')
+        for contract in self:
+            if contract.parent_id:
+                logger.info("Contract " + str(contract.id) + " contract sub.")
+                wf_service.trg_validate(
+                    self.env.user.id, self._name, contract.parent_id.id,
+                    'new_contract_validated', self.env.cr)
+        return True
+
+    @api.multi
+    def contract_cancelled(self):
+        """ Project state is no more relevant when contract is cancelled. """
+        res = super(recurring_contract, self).contract_cancelled()
+        self.write({'project_state': False})
+        return res
+
+    @api.multi
+    def contract_terminated(self):
+        """ Project state is no more relevant when contract is terminated.
+        We also put the person who terminated the contract as follower. """
+        res = super(recurring_contract, self).contract_terminated()
+        self.write({'project_state': False,
+                    'sds_uid': self.env.user.id})
+        return res
+
+    ##########################################################################
+    #                             PRIVATE METHODS                            #
+    ##########################################################################
+    def _define_parent_id(self, correspondant_id):

```

Comment:
Not necessary in private methods (only called from this class, api.multi is by default)